oximeter/impl/src/lib.rs

```diff
@@ -0,0 +1,45 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2024 Oxide Computer Company
+
+pub use oximeter_macro_impl::*;
+
+// Export the current crate as `oximeter`. The macros defined in `oximeter-macro-impl` generate
+// code referring to symbols like `oximeter::traits::Target`. In consumers of this crate, that's
+// fine, but internally there _is_ no crate named `oximeter`, it's just `self` or `crate`.
+//
+// See https://github.com/rust-lang/rust/pull/55275 for the PR introducing this fix, which links to
+// lots of related issues and discussion.
+extern crate self as oximeter;
+
+pub mod histogram;
+pub mod schema;
+pub mod test_util;
+pub mod traits;
+pub mod types;
+
+pub use schema::FieldSchema;
+pub use schema::TimeseriesName;
+pub use schema::TimeseriesSchema;
+pub use traits::Metric;
+pub use traits::Producer;
+pub use traits::Target;
+pub use types::Datum;
+pub use types::DatumType;
+pub use types::Field;
+pub use types::FieldType;
+pub use types::FieldValue;
+pub use types::Measurement;
+pub use types::MetricsError;
+pub use types::Sample;
+
+/// Construct the timeseries name for a Target and Metric.
+pub fn timeseries_name<T, M>(target: &T, metric: &M) -> TimeseriesName
+where
+    T: Target,
+    M: Metric,
+{
+    TimeseriesName(format!("{}:{}", target.name(), metric.name()))
```

Comment:
Maybe. the `Target::name` and `Metric::name` methods all return the snake_case name of the struct right now (assuming they're derived, which they all are, I checked). So since `:` isn't valid in a Rust identifier, we can't have the names contain that character.

That said, it's definitely _possible_, in that folks could implement the traits themselves and create invalid names. I don't have a strong preference, so would be happy to add that check if you think it's worthwhile.

Reply:
I guess there are three options:

1. Leave as-is
2. Change this to `try_from()` and unwrap the result
3. Change this to `try_from()` and return an error

I think 1 and 2 are almost the same: derived implementations will be fine, an incorrect manual implementation would panic here in 2, and an incorrect manual implementation would (probably?) panic somewhere else later? 3 would be the safest but might also be annoying since this function is currently infallible.

If 3 is super easy I'd do that, but if it's annoying I might go with option 2, and make it something like `.try_from().expect("illegal name: implementations should be derived")` or whatever to make it clear that the only reason it's failing is because someone went out of their way to do it wrong.

Reply:
I made this fallible in d358d61c8